File: GdPicture/WinForm/vb.net/TIFF to PDF-OCR/Form1.vb
Option Strict On
Option Explicit On

Imports System.IO
Imports GdPicture14

Public Class Form1

    Private ReadOnly _nativePdf As New GdPicturePDF
    Private _cancellationPending As Boolean


    Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
        Dim oLicenseManager As New GdPicture14.LicenseManager()
        oLicenseManager.RegisterKEY("XXX") 'Please replace XXX by a valid demo or commercial key.

        txtDictsPath.Text = oLicenseManager.GetRedistPath() + "OCR\"

        AddHandler _nativePdf.OcrPagesProgress, AddressOf Me.OcrPagesProgress
        AddHandler _nativePdf.BeforePageOcr, AddressOf Me.BeforePageOcr
        AddHandler _nativePdf.OcrPagesDone, AddressOf Me.OcrPagesDone
    End Sub


    Private Sub txtDictsPath_TextChanged(sender As Object, e As EventArgs) Handles txtDictsPath.TextChanged
        If Directory.Exists(txtDictsPath.Text) Then
            Dim d As New DirectoryInfo(txtDictsPath.Text)
            Dim files As FileInfo() = d.GetFiles("*.traineddata")
            For Each file As FileInfo In files
                cmbLang.Items.Add(Path.GetFileNameWithoutExtension(file.Name))
            Next
        End If
    End Sub


    Private Sub BeforePageOcr(ByVal pageNo As Integer, ByRef cancel As Boolean)
        'nothing to do
    End Sub


    Private Sub OcrPagesProgress(ByVal status As GdPictureStatus, ByVal pageNo As Integer, ByVal processed As Integer, ByVal count As Integer, ByRef cancel As Boolean)
        ProgressBar1.Maximum = count
        ProgressBar1.Value = processed
        If status <> GdPictureStatus.OK Then
            If Not _cancellationPending Then
                If MessageBox.Show("An error occured on page " + pageNo.ToString() + ". Do you want to cancel the process? Status: " + status.ToString(), "error", MessageBoxButtons.YesNo, MessageBoxIcon.Stop) = DialogResult.Yes Then
                    cancel = True
                    _cancellationPending = True
                End If
            End If

        End If
    End Sub


    Private Sub OcrPagesDone(ByVal status As GdPictureStatus)
        If status = GdPictureStatus.OK Then
            Dim outputFilePath As String = Path.ChangeExtension(txtInputFile.Text, "pdf")
            status = _nativePdf.SaveToFile(outputFilePath, True)
            If status <> GdPictureStatus.OK Then
                MessageBox.Show("Can't save file " + outputFilePath + " . Error: " + status.ToString())
            Else
                Process.Start(outputFilePath)
            End If
            MessageBox.Show("Done! Please check file " + outputFilePath)
        Else
            MessageBox.Show("An error occured. Status: " + status.ToString())
        End If

        _nativePdf.CloseDocument()
        UnlockUi()
    End Sub


    Private Sub LockUi()
        Me.Enabled = False
        ProgressBar1.Visible = True
        ProgressBar1.Value = 0
    End Sub


    Private Sub UnlockUi()
        Me.Enabled = True
        ProgressBar1.Visible = False
    End Sub


    Private Sub btStartOCR_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btStartOCR.Click
        If Not String.IsNullOrEmpty(txtInputFile.Text) Then
            If Not String.IsNullOrEmpty(cmbLang.Text) Then

                Using gdpictureImaging As New GdPictureImaging
                    Dim imageID As Integer = gdpictureImaging.CreateGdPictureImageFromFile(txtInputFile.Text)
                    If gdpictureImaging.GetStat = GdPictureStatus.OK Then
                        Dim resolution As Single = Math.Max(200, gdpictureImaging.GetVerticalResolution(imageID))
                        _nativePdf.NewPDF(PdfConformance.PDF_A_1b)
                        For i As Integer = 1 To gdpictureImaging.GetPageCount(imageID)
                            If gdpictureImaging.SelectPage(imageID, i) = GdPictureStatus.OK Then
                                _nativePdf.AddImageFromGdPictureImage(imageID, False, True)
                            End If
                        Next
                        gdpictureImaging.ReleaseGdPictureImage(imageID)

                        LockUi()
                        _cancellationPending = False

                        Dim threadCount As Integer
                        If chkMultithread.Checked Then
                            threadCount = 0 '0 means automatic number of thread computation
                        Else
                            threadCount = 1
                        End If

                        Dim oCRMode As OCRMode
                        If chkFavorAccuracy.Checked Then
                            oCRMode = OCRMode.FavorAccuracy 'automatic
                        Else
                            oCRMode = OCRMode.FavorSpeed 'faster
                        End If

                        _nativePdf.OcrPages("*", threadCount, cmbLang.Text, txtDictsPath.Text, "", resolution, oCRMode, 0, False)
                    Else
                        MessageBox.Show("Can't open file: " + txtInputFile.Text + " status: " + gdpictureImaging.GetStat().ToString())
                    End If

                End Using

            Else
                MessageBox.Show("Please, select a language to recognize.")
            End If
        Else
            MessageBox.Show("Please, select a file to convert.")
        End If
    End Sub


    Private Sub btBrowsInput_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btBrowsInput.Click
        OpenFileDialog1.FileName = ""
        OpenFileDialog1.Filter = "Tiff Image Files(*.TIF;*.TIFF)|*.TIF;*.TIFF|All files (*.*)|*.*"
        OpenFileDialog1.Multiselect = False
        If OpenFileDialog1.ShowDialog() = DialogResult.OK Then
            txtInputFile.Text = OpenFileDialog1.FileName
        End If
    End Sub


    Private Sub btBrowsOCRDict_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btBrowsOCRDict.Click
        If FolderBrowserDialog1.ShowDialog = DialogResult.OK Then
            txtDictsPath.Text = FolderBrowserDialog1.SelectedPath
        End If
    End Sub
End Class

File: GdPicture/WinForm/C#/Image to PDF-OCR/Form1.cs
using System.Diagnostics;
using System.IO;
using System.Windows.Forms;
using GdPicture14;

namespace Image2PDFOCR
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private readonly GdPicturePDF _nativePdf = new GdPicturePDF();
        private bool _cancellationPending;

        private void Form1_Load(object sender, System.EventArgs e)
        {
            LicenseManager oLicenseManager = new LicenseManager(); //Go to http://www.gdpicture.com/download-gdpicture/ to get a 1 month trial key unlocking all features of the toolkit.
            oLicenseManager.RegisterKEY("XXXX"); //Please, replace XXXX by a valid demo or commercial license key.

            txtDictsPath.Text = oLicenseManager.GetRedistPath() + "OCR\\";

            _nativePdf.OcrPagesProgress += this.OcrPagesProgress;
            _nativePdf.BeforePageOcr += this.BeforePageOcr;
            _nativePdf.OcrPagesDone += this.OcrPagesDone;
        }

        private void txtDictsPath_TextChanged(object sender, System.EventArgs e)
        {
            if (Directory.Exists(txtDictsPath.Text))
            {
                DirectoryInfo d = new DirectoryInfo(txtDictsPath.Text);
                FileInfo[] files = d.GetFiles("*.traineddata");
                foreach (FileInfo file in files)
                {
                    cmbLang.Items.Add(Path.GetFileNameWithoutExtension(file.Name));
                }
            }
        }


        private void BeforePageOcr(int pageNo, ref bool cancel)
        {
            //anything to do
        }


        private void OcrPagesProgress(GdPictureStatus status, int pageNo, int processed, int count, ref bool cancel)
        {
            ProgressBar1.Maximum = count;
            ProgressBar1.Value = processed;
            if (status != GdPictureStatus.OK)
            {
                if (!_cancellationPending)
                {
                    if (MessageBox.Show("An error occured on page " + pageNo + ". Do you want to cancel the process? Status: " + status, "error", MessageBoxButtons.YesNo, MessageBoxIcon.Stop) == DialogResult.Yes)
                    {
                        cancel = true;
                        _cancellationPending = true;
                    }
                }
            }
        }


        private void OcrPagesDone(GdPictureStatus status)
        {
            if (status == GdPictureStatus.OK)
            {
                string outputFilePath = Path.ChangeExtension(txtInputFile.Text, "pdf");

                status = _nativePdf.SaveToFile(outputFilePath);

                if (status != GdPictureStatus.OK)
                {
                    MessageBox.Show("Can't save file " + outputFilePath + " . Error: " + status);
                }
                else
                {
                    Process.Start(outputFilePath);
                }

                MessageBox.Show("Done! Please check file " + outputFilePath);
            }
            else
            {
                MessageBox.Show("An error occured. Status: " + status);
            }

            _nativePdf.CloseDocument();
            UnlockUi();
        }


        private void LockUi()
        {
            this.Enabled = false;
            ProgressBar1.Visible = true;
            ProgressBar1.Value = 0;
        }


        private void UnlockUi()
        {
            this.Enabled = true;
            ProgressBar1.Visible = false;
        }


        private void btStartOCR_Click(object sender, System.EventArgs e)
        {
            if (!string.IsNullOrEmpty(txtInputFile.Text))
            {
                if (!string.IsNullOrEmpty(cmbLang.Text))
                {
                    using (GdPictureImaging gdpictureImaging = new GdPictureImaging())
                    {
                        int imageID = gdpictureImaging.CreateGdPictureImageFromFile(txtInputFile.Text);
                        if (gdpictureImaging.GetStat() == GdPictureStatus.OK)
                        {
                            float resolution = System.Math.Max(200, gdpictureImaging.GetVerticalResolution(imageID));
                            _nativePdf.NewPDF(PdfConformance.PDF_A_1b);

                            for (int i = 1; i <= gdpictureImaging.GetPageCount(imageID); i++)
                            {
                                if (gdpictureImaging.SelectPage(imageID, i) == GdPictureStatus.OK)
                                {
                                    _nativePdf.AddImageFromGdPictureImage(imageID, false, true);
                                }
                            }
                            gdpictureImaging.ReleaseGdPictureImage(imageID);

                            LockUi();
                            _cancellationPending = false;

                            int threadCount;
                            if (chkMultithread.Checked)
                            {
                                threadCount = 0;
                                //0 means automatic number of thread computation
                            }
                            else
                            {
                                threadCount = 1;
                            }

                            _nativePdf.OcrPages("*", threadCount, cmbLang.Text, txtDictsPath.Text, "", resolution, chkFavorAccuracy.Checked ? OCRMode.FavorAccuracy : OCRMode.FavorSpeed, 0, false);
                        }
                        else
                        {
                            MessageBox.Show("Can't open file: " + txtInputFile.Text + " status: " + gdpictureImaging.GetStat().ToString());
                        }
                    }
                }
                else
                {
                    MessageBox.Show("Please, select a language to recognize.");
                }
            }
            else
            {
                MessageBox.Show("Please, select a file to convert.");
            }
        }


        private void btBrowsInput_Click(object sender, System.EventArgs e)
        {
            using (OpenFileDialog openFileDialog1 = new OpenFileDialog())
            {
                openFileDialog1.FileName = "";
                openFileDialog1.Filter = "Image File|*.*";
                openFileDialog1.Multiselect = false;
                if (openFileDialog1.ShowDialog() == DialogResult.OK)
                {
                    txtInputFile.Text = openFileDialog1.FileName;
                }
            }
        }


        private void btBrowsOCRDict_Click(object sender, System.EventArgs e)
        {
            using (FolderBrowserDialog folderBrowserDialog1 = new FolderBrowserDialog())
            {
                if (folderBrowserDialog1.ShowDialog() == DialogResult.OK)
                {
                    txtDictsPath.Text = folderBrowserDialog1.SelectedPath;
                }
            }
        }
    }
}
